fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/matching/BaseMatcher.java

```diff
@@ -0,0 +1,404 @@
+/*
+ * BaseMatcher.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp.matching;
+
+import com.apple.foundationdb.record.query.plan.temp.AliasMap;
+import com.apple.foundationdb.record.query.plan.temp.CorrelationIdentifier;
+import com.apple.foundationdb.record.query.plan.temp.EnumeratingIterable;
+import com.apple.foundationdb.record.query.plan.temp.TransitiveClosure;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.apple.foundationdb.record.query.plan.temp.TopologicalSort.anyTopologicalOrderPermutation;
+import static com.apple.foundationdb.record.query.plan.temp.TopologicalSort.topologicalOrderPermutations;
+
+/**
+ * Abstract base class for all matchers. Matching is an algorithm that <em>matches</em> two collections of things of type
+ * {@code T} resulting in a bijective mapping (injective and surjective) between the elements of the respective
```

Comment:
I'm not sure how helpful explaining "bijective" that way is: if you know "injective" and "surjective," you know it. I would spell it out: every element of this collection has a corresponding element of the other collection and vice versa.